internal/plugin/server.go

```diff
@@ -42,68 +41,47 @@ import (
 	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
 )
 
-// Constants for use by the 'volume-mounts' device list strategy
 const (
+	deviceListEnvVar                          = "NVIDIA_VISIBLE_DEVICES"
 	deviceListAsVolumeMountsHostPath          = "/dev/null"
 	deviceListAsVolumeMountsContainerPathRoot = "/var/run/nvidia-container-devices"
 )
 
-// NvidiaDevicePlugin implements the Kubernetes device plugin API
-type NvidiaDevicePlugin struct {
+// nvidiaDevicePlugin implements the Kubernetes device plugin API
+type nvidiaDevicePlugin struct {
 	rm                   rm.ResourceManager
 	config               *spec.Config
-	deviceListEnvVar     string
 	deviceListStrategies spec.DeviceListStrategies
-	socket               string
-	kubeletSocket        string
 
 	cdiHandler          cdi.Interface
-	cdiEnabled          bool
 	cdiAnnotationPrefix string
 
+	socket string
 	server *grpc.Server
 	health chan *rm.Device
 	stop   chan interface{}
 
-	mpsDaemon   *mps.Daemon
-	mpsHostRoot mps.Root
+	mps mpsOptions
 }
 
-// NewNvidiaDevicePlugin returns an initialized NvidiaDevicePlugin
-func NewNvidiaDevicePlugin(config *spec.Config, kubeletSocket string, resourceManager rm.ResourceManager, cdiHandler cdi.Interface) (*NvidiaDevicePlugin, error) {
-	_, name := resourceManager.Resource().Split()
-
-	deviceListStrategies, _ := spec.NewDeviceListStrategies(*config.Flags.Plugin.DeviceListStrategy)
-
-	pluginName := "nvidia-" + name
-	pluginPath := filepath.Join(pluginapi.DevicePluginPath, pluginName)
-
-	var mpsDaemon *mps.Daemon
-	var mpsHostRoot mps.Root
-	if config.Sharing.SharingStrategy() == spec.SharingStrategyMPS {
-		// TODO: It might make sense to pull this logic into a resource manager.
-		for _, device := range resourceManager.Devices() {
-			if device.IsMigDevice() {
-				return nil, errors.New("sharing using MPS is not supported for MIG devices")
-			}
-		}
-		mpsDaemon = mps.NewDaemon(resourceManager, mps.ContainerRoot)
-		mpsHostRoot = mps.Root(*config.Flags.CommandLineFlags.MpsRoot)
+// devicePluginForResource creates a device plugin for the specified resource.
+func (o *options) devicePluginForResource(resourceManager rm.ResourceManager) (Interface, error) {
+	mpsOptions, err := o.getMPSOptionsIfConfigured(resourceManager)
```

Comment:
actually -- with it refactored like this, I think we can drop the `IfConfigured` since now you are getting back a set of options, which will indicate if it is enabled or not inside of it.